### PR TITLE
Alter crossover operators to hold children between successive pipeline calls

### PR DESCRIPTION
**Pull request by lukepmccombs**
Fixes an issue mentioned in #244 

Alters crossover operators to return a function that performs the crossover operation instead. This lets them capture locals, in this case a deque for children that persists between successive pipeline calls. This allows children that should be released to the population, but weren't due to budgeting in successive operators, to be properly represented.

### CODE REVIEW

File: leap_ec/ops.py
Comment (SigmaX) [inline leap_ec/ops.py:424]: So, the deque is never emptied unless the user explicitly empties it externally?

This will cause problems with long runs/large genomes, won't it, tantamount to a memory leak?

Not sure how I feel about this as default behavior.

**Review by SigmaX — CHANGES_REQUESTED** [review]
Thanks for the PR!

 - [x] I'm concerned that the `deque` just keeps filling up by default unless the user explicitly adds external logic to empty it.  As default behavior, this could surprise users with what effectively acts as a memory leak.  Perhaps we could solve this by adding a binary flag to toggle the `deque` behavior, and have it default to `False`?
 - [ ] Additional tests are needed to cover the new behavior
 - [ ] A few tests are still failing when I run them:

image

File: leap_ec/ops.py
Comment (lukepmccombs) [reply leap_ec/ops.py:436]: I've written an alternate version of the function in #244 that is a bit more understandable in the context of how it had been done before.

File: leap_ec/ops.py
Comment (lukepmccombs) [reply leap_ec/ops.py:424]: Only ever 1-2 individuals are in it at a time, and the second only briefly. I am an advocate for having a parameter to override the behavior, though.

File: leap_ec/ops.py
Comment (lukepmccombs) [reply leap_ec/ops.py:355]: I agree, I'll get on that.

**Comment by lukepmccombs** [comment]
I'm uncertain why those tests are failing, since they weren't failing for me locally. I'll take a look.

File: leap_ec/ops.py
Comment (SigmaX) [inline leap_ec/ops.py:392]: I think this test will pass even if `persist_children=False`?  So we'll need to modify it to specifically target what is dfferent about the `persist_children` behavior.

**Comment by SigmaX** [comment]
@lukepmccombs You're right—from the beginning we've had a tendency to use a mix of closures (regular functions), currying, and classes for operators.  Preferring `curry` for operators that have not required arguments allowed us to start standardizing on a zero-parentheses syntax for them... but your revelations here about how curried functions don't persist now throw a wrench in that convention!  @markcoletti, any thoughts on this problem?

———

First, making sure I understand: the observation you made in #244 is that when our crossover operators are used in a _steady-state population model_ specifically, then because we are generating a single offspring at a time in each evolutionary step, we end up losing the functionality of generating a pair of two offspring for each pair of parents.

So the original operators work as designed—one instance will correctly generate two complementary offspring when you call it twice.  The problem is that we are only calling the pipeline once in each step for steady-state algorithms, so the second offspring gets created but never yielded (thus inflating our birth count).

———
In that case, I think the only thing that is holding up a merge is the unit tests don't quite target the new behavior precisely yet...

File: leap_ec/ops.py
Comment (lukepmccombs) [reply leap_ec/ops.py:392]: Nope, just tested it! I'll try to make it clearer, but the important part is that `op(select)` is being called twice. This causes the issue where the second yielded individual is never consumed, since it restarts the loop over again.

**Comment by lukepmccombs** [comment]
I've expanded the unit tests to have both cases of `persist_individuals` with a hopefully better accompanying explanation.

**Comment by markcoletti** [comment]
> 

I *think* most of the legacy pipeline operators are ok with being reinstantiated because they're not stateful in the sense that once they're pulsed they essentially reset, in a sense.  The problem is that crossover inherently maintains state between invocations where it buffers the second individual.  So, this may be a situation unique to such operators that yield more than one thing.  @SigmaX , what do you think?

**Comment by markcoletti** [comment]
> @lukepmccombs That proposed convention sounds like a good idea to me. If @markcoletti agrees, I would be okay with replacing our current crossover ops with classes (and adding a line to their docstring explaining that they are classes because they have some state).
> 
> If we want to do that, I'd suggest we go ahead and do it as part of this issue/PR, and get it over with.

I agree! 👍

**Review by SigmaX — APPROVED** [review]
Test & looks good.  All that's missing is a CHANGELOG entry—we can add that after merging.

**Pull request merged**
